rtsp.go

```diff
@@ -239,71 +259,126 @@ func (rc *rtspCamera) initH264(tracks media.Medias, baseURL *url.URL) (err error
 	}
 
 	// if SPS and PPS are present into the SDP, send them to the decoder
-	if format.SPS != nil {
-		rc.rawDecoder.decode(format.SPS)
+	if f.SPS != nil {
+		rc.rawDecoder.decode(f.SPS) // nolint:errcheck
 	} else {
 		rc.logger.Warn("no SPS found in H264 format")
 	}
-	if format.PPS != nil {
-		rc.rawDecoder.decode(format.PPS)
+	if f.PPS != nil {
+		rc.rawDecoder.decode(f.PPS) // nolint:errcheck
 	} else {
 		rc.logger.Warn("no PPS found in H264 format")
 	}
 
-	// On packet retreival, turn it into an image, and store it in shared memory
-	rc.client.OnPacketRTP(track, format, func(pkt *rtp.Packet) {
-		// extract access units from RTP packets
-		au, _, err := rtpDec.Decode(pkt)
+	var waitingForIframeLogged bool
+	iFrameReceived := false
+	storeImage := func(pkt *rtp.Packet) {
+		au, err := rtpDec.Decode(pkt)
 		if err != nil {
 			if err != rtph264.ErrNonStartingPacketAndNoPrevious && err != rtph264.ErrMorePacketsNeeded {
-				rc.logger.Errorf("error decoding(1) h264 rstp stream %v", err)
+				rc.logger.Debugf("error decoding(1) h264 rstp stream %v", err)
 			}
 			return
 		}
 
-		for _, nalu := range au {
+		if !iFrameReceived {
+			if !h264.IDRPresent(au) {
+				if !waitingForIframeLogged {
+					rc.logger.Debug("waiting for I-frame")
+					waitingForIframeLogged = true
+				}
+				return
+			}
+			iFrameReceived = true
+			rc.logger.Debug("got I-frame")
+		}
 
-			if len(nalu) < 20 {
-				// TODO(ERH): this is probably wrong, but fixes a spam issue with "no frame!"
-				continue
+		for _, nalu := range au {
+			image, err := rc.rawDecoder.decode(nalu)
+			if err != nil {
+				rc.logger.Error("error decoding(2) h264 rtsp stream  %v", err)
+				return
+			}
+			if image != nil {
+				rc.latestFrame.Store(&image)
 			}
+		}
+	}
 
-			// convert NALUs into RGBA frames
-			lastImage, err := rc.rawDecoder.decode(nalu)
+	onPacketRTP := func(pkt *rtp.Packet) {
```

Comment:
If passthrough is not enabled we only want to execute the `storeImage` callback to maintain current behavior